FreeBSD/usr/src/share/man/man4/sctp.4
.\" Copyright (c) 2006, Randall Stewart.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: releng/11.1/share/man/man4/sctp.4 294114 2016-01-16 00:37:27Z imp $
.\"
.Dd December 15, 2006
.Dt SCTP 4
.Os
.Sh NAME
.Nm sctp
.Nd Internet Stream Control Transmission Protocol
.Sh SYNOPSIS
.In sys/types.h
.In sys/socket.h
.In netinet/sctp.h
.Ft int
.Fn socket AF_INET SOCK_STREAM IPPROTO_SCTP
.Ft int
.Fn socket AF_INET SOCK_SEQPACKET IPPROTO_SCTP
.Sh DESCRIPTION
The
.Tn SCTP
protocol provides reliable, flow-controlled, two-way
transmission of data.
It is a message oriented protocol and can
support the
.Dv SOCK_STREAM
and
.Dv SOCK_SEQPACKET
abstractions.
.Tn SCTP
uses the standard
Internet address format and, in addition, provides a per-host
collection of
.Dq "port addresses" .
Thus, each address is composed of an Internet address specifying
the host and network, with a specific
.Tn SCTP
port on the host identifying the peer entity.
.Pp
There are two models of programming in SCTP.
The first uses the
.Dv SOCK_STREAM
abstraction.
In this abstraction sockets utilizing the
.Tn SCTP
protocol are either
.Dq active
or
.Dq passive .
Active sockets initiate connections to passive
sockets.
By default,
.Tn SCTP
sockets are created active; to create a
passive socket, the
.Xr listen 2
system call must be used after binding the socket with the
.Xr bind 2
or
.Xr sctp_bindx 3
system calls.
Only passive sockets may use the
.Xr accept 2
call to accept incoming connections.
Only active sockets may use the
.Xr connect 2
call to initiate connections.
.Pp
The other abstraction
.Dv SOCK_SEQPACKET
provides a
.Dq connectionless
mode of operation in that the user may send to an address
(using any of the valid send calls that carry a
socket address) and an association will be setup
implicitly by the underlying
.Tn SCTP
transport stack.
This abstraction is the only one capable of sending data on the
third leg of the four-way handshake.
A user must still call
.Xr listen 2
to allow the socket to accept connections.
Calling
.Xr listen 2
however does not restrict the user from still initiating
implicit connections to other peers.
.Pp
The
.Tn SCTP
protocol directly supports multi-homing.
So when binding a socket with the
.Dq wildcard
address
.Dv INADDR_ANY ,
the
.Tn SCTP
stack will inform the peer about all of the local addresses
that are deemed in scope of the peer.
The peer will then possibly have multiple paths to reach the local host.
.Pp
The
.Tn SCTP
transport protocol is also multi-streamed.
Multi-streaming refers to the ability to send sub-ordered flows of
messages.
A user performs this by specifying a specific stream in one of the
extended send calls such as the
.Xr sctp_send 3
function call.
Sending messages on different streams will allow parallel delivery
of data i.e., a message loss in stream 1 will not block the delivery
of messages sent in stream 2.
.Pp
The
.Tn SCTP
transport protocol also provides a unordered service as well.
The unordered service allows a message to be sent and delivered
with no regard to the ordering of any other message.
.Ss Extensions
The FreeBSD implementation of
.Tn SCTP
also supports the following extensions:
.Bl -hang -width indent
.It "sctp partial reliability"
This extension allows one to have message be skipped and
not delivered based on some user specified parameters.
.It "sctp dynamic addressing"
This extension allows addresses to be added and deleted
dynamically from an existing association.
.It "sctp authentication"
This extension allows the user to authenticate specific
peer chunks (including data) to validate that the peer
who sent the message is in fact the peer who setup the
association.
A shared key option is also provided for
so that two stacks can pre-share keys.
.It "packet drop"
Some routers support a special satellite protocol that
will report losses due to corruption.
This allows retransmissions without subsequent loss in bandwidth
utilization.
.It "stream reset"
This extension allows a user on either side to reset the
stream sequence numbers used by any or all streams.
.El
.Pp
.Tn SCTP
supports a number of socket options which can be set with
.Xr setsockopt 2
and tested with
.Xr getsockopt 2
or
.Xr sctp_opt_info 3 :
.Bl -tag -width ".Dv SCTP_SET_PEER_PRIMARY_ADDR"
.It Dv SCTP_NODELAY
Under most circumstances,
.Tn SCTP
sends data when it is presented; when outstanding data has not
yet been acknowledged, it gathers small amounts of output to be
sent in a single packet once an acknowledgement is received.
For some clients, such as window systems that send a stream of
mouse events which receive no replies, this packetization may
cause significant delays.
The boolean option
.Dv SCTP_NODELAY
defeats this algorithm.
.It Dv SCTP_RTOINFO
This option returns specific information about an associations
.Dq "Retransmission Time Out" .
It can also be used to change the default values.
.It Dv SCTP_ASSOCINFO
This option returns specific information about the requested
association.
.It Dv SCTP_INITMSG
This option allows you to get or set the default sending
parameters when an association is implicitly setup.
It allows you to change such things as the maximum number of
streams allowed inbound and the number of streams requested
of the peer.
.It Dv SCTP_AUTOCLOSE
For the one-to-many model
.Dv ( SOCK_SEQPACKET )
associations are setup implicitly.
This option allows the user to specify a default number of idle
seconds to allow the association be maintained.
After the idle timer (where no user message have been sent or have
been received from the peer) the association will be gracefully
closed.
The default for this value is 0, or unlimited (i.e., no automatic
close).
.It Dv SCTP_SET_PEER_PRIMARY_ADDR
The dynamic address extension allows a peer to also request a
particular address of its be made into the primary address.
This option allows the caller to make such a request to a peer.
Note that if the peer does not also support the dynamic address
extension, this call will fail.
Note the caller must provide a valid local address that the peer has
been told about during association setup or dynamically.
.It Dv SCTP_PRIMARY_ADDR
This option allows the setting of the primary address
that the caller wishes to send to.
The caller provides the address of a peer that is to be made primary.
.It Dv SCTP_ADAPTATION_LAYER
The dynamic address extension also allows a user to
pass a 32 bit opaque value upon association setup.
This option allows a user to set or get this value.
.It Dv SCTP_DISABLE_FRAGMENTS
By default
.Tn SCTP
will fragment user messages into multiple pieces that
will fit on the network and then later, upon reception, reassemble
the pieces into a single user message.
If this option is enabled instead, any send that exceeds the path
maximum transfer unit (P-MTU) will fail and the message will NOT be
sent.
.It Dv SCTP_PEER_ADDR_PARAMS
This option will allow a user to set or get specific
peer address parameters.
.It Dv SCTP_DEFAULT_SEND_PARAM
When a user does not use one of the extended send
calls (e.g.,
.Xr sctp_sendmsg 3 )
a set of default values apply to each send.
These values include things like the stream number to send
to as well as the per-protocol id.
This option lets a caller both get and set these values.
If the user changes these default values, then these new values will
be used as the default whenever no information is provided by the
sender (i.e., the non-extended API is used).
.It Dv SCTP_EVENTS
.Tn SCTP
has non-data events that it can communicate
to its application.
By default these are all disabled since they arrive in the data path
with a special flag
.Dv MSG_NOTIFICATION
set upon the received message.
This option lets a caller
both get what events are current being received
as well as set different events that they may be interested
in receiving.
.It Dv SCTP_I_WANT_MAPPED_V4_ADDR
.Tn SCTP
supports both IPV4 and IPV6.
An association may span both IPV4 and IPV6 addresses since
.Tn SCTP
is multi-homed.
By default, when opening an IPV6 socket, when
data arrives on the socket from a peer's
V4 address the V4 address  will be presented with an address family
of AF_INET.
If this is undesirable, then this option
can be enabled which will then convert all V4 addresses
into mapped V6 representations.
.It Dv SCTP_MAXSEG
By default
.Tn SCTP
chooses its message fragmentation point
based upon the smallest P-MTU of the peer.
This option lets the caller set it to a smaller value.
Note that while the user can change this value, if the P-MTU
is smaller than the value set by the user, then the P-MTU
value will override any user setting.
.It Dv SCTP_DELAYED_ACK_TIME
This option lets the user both set and get the
delayed ack time (in milliseconds) that
.Tn SCTP
is using.
The default is 200 milliseconds.
.It Dv SCTP_PARTIAL_DELIVERY_POINT
.Tn SCTP
at times may need to start delivery of a
very large message before the entire message has
arrived.
By default SCTP waits until the incoming
message is larger than one fourth of the receive
buffer.
This option allows the stacks value
to be overridden with a smaller value.
.It Dv SCTP_FRAGMENT_INTERLEAVE
.Tn SCTP
at times will start partial delivery (as mentioned above).
In the normal case successive reads will continue to return
the rest of the message, blocking if needed, until all of
that message is read.
However this means other messages may have arrived and be ready
for delivery and be blocked behind the message being partially
delivered.
If this option is enabled, when a partial delivery
message has no more data to be received, then a subsequent
read may return a different message that is ready for delivery.
By default this option is off since the user must be using the
extended API's to be able to tell the difference between
messages (via the stream and stream sequence number).
.It Dv SCTP_AUTH_CHUNK
By default only the dynamic addressing chunks are
authenticated.
This option lets a user request an
additional chunk be authenticated as well.
Note that successive calls to this option will work and continue
to add more chunks that require authentication.
Note that this option only effects future associations and
not existing ones.
.It Dv SCTP_AUTH_KEY
This option allows a user to specify a shared
key that can be later used to authenticate
a peer.
.It Dv SCTP_HMAC_IDENT
This option will let you get or set the list of
HMAC algorithms used to authenticate peers.
Note that the HMAC values are in priority order where
the first HMAC identifier is the most preferred
and the last is the least preferred.
.It Dv SCTP_AUTH_ACTIVE_KEY
This option allows you to make a key active for
the generation of authentication information.
Note that the peer must have the same key or else the
data will be discarded.
.It Dv SCTP_AUTH_DELETE_KEY
This option allows you to delete an old key.
.It Dv SCTP_USE_EXT_RECVINFO
The sockets api document allows an extended
send/receive information structure to be used.
The extended structure includes additional fields
related to the next message to be received (after the
current receive completes) if such information is known.
By default the system will not pass this information.
This option allows the user to request this information.
.It Dv SCTP_AUTO_ASCONF
By default when bound to all address and the system administrator has
enables automatic dynamic addresses, the
.Tn SCTP
stack will automatically generate address changes into add and
delete requests to any peers by setting this option to
true.
This option allows an endpoint to disable that behavior.
.It Dv SCTP_MAXBURST
By default
.Tn SCTP
implements micro-burst control so that as the congestion window
opens up no large burst of packets can be generated.
The default burst limit is four.
This option lets the user change this value.
.It Dv SCTP_CONTEXT
Many sctp extended calls have a context field.
The context field is a 32 bit opaque value that will be returned in
send failures.
This option lets the caller set the default
context value to use when none is provided by the user.
.It Dv SCTP_EXPLICIT_EOR
By default, a single send is a complete message.
.Tn SCTP
generates an implied record boundary.
If this option is enabled, then all sends are part of the same message
until the user indicates an end of record with the
special flag
.Dv SCTP_EOR
passed in the sctp_sndrcvinfo flags field.
This effectively makes all sends part of the same message
until the user specifies differently.
This means that a caller must NOT change the stream number until
after the
.Dv SCTP_EOR
is passed to
.Tn SCTP
else an error will be returned.
.It Dv SCTP_STATUS
This option is a read-only option that returns
various status information about the specified association.
.It Dv SCTP_GET_PEER_ADDR_INFO
This read-only option returns information about a peer
address.
.It Dv SCTP_PEER_AUTH_CHUNKS
This read-only option returns a list of the chunks
the peer requires to be authenticated.
.It Dv SCTP_LOCAL_AUTH_CHUNKS
This read-only option returns a list of the locally
required chunks that must be authenticated.
.It Dv SCTP_RESET_STREAMS
This socket option is used to cause a stream sequence
number or all stream sequence numbers to be reset.
Note that the peer
.Tn SCTP
endpoint must also support the stream reset extension
as well.
.El
.Sh SEE ALSO
.Xr accept 2 ,
.Xr bind 2 ,
.Xr connect 2 ,
.Xr listen 2 ,
.Xr sctp_bindx 3 ,
.Xr sctp_connectx 3 ,
.Xr sctp_opt_info 3 ,
.Xr sctp_recvmsg 3 ,
.Xr sctp_sendmsg 3


FreeBSD/usr/src/usr.bin/systat/sctp.c
/*-
 * Copyright (c) 2015
 * The Regents of the University of California.  All rights reserved.
 * Michael Tuexen.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/11.1/usr.bin/systat/sctp.c 289873 2015-10-24 09:34:40Z tuexen $");

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sysctl.h>

#include <netinet/sctp.h>

#include <stdlib.h>
#include <string.h>

#include "systat.h"
#include "extern.h"
#include "mode.h"

static struct sctpstat curstat, initstat, oldstat;

/*-
--0         1         2         3         4         5         6         7
--0123456789012345678901234567890123456789012345678901234567890123456789012345
00             SCTP Associations                     SCTP Packets
01999999999999 associations initiated   999999999999 packets sent
02999999999999 associations accepted    999999999999 packets received
03999999999999 associations established 999999999999 - out of the blue
04999999999999 associations restarted   999999999999 - bad vtag
05999999999999 associations terminated  999999999999 - bad crc32c
06999999999999 associations aborted
07
08             SCTP Timers                           SCTP Chunks
09999999999999 init timeouts            999999999999 control chunks sent
10999999999999 cookie timeouts          999999999999 data chunks sent
11999999999999 data timeouts            999999999999 - ordered
12999999999999 delayed sack timeouts    999999999999 - unordered
13999999999999 shutdown timeouts        999999999999 control chunks received
14999999999999 shutdown-ack timeouts    999999999999 data chunks received
15999999999999 shutdown guard timeouts  999999999999 - ordered
16999999999999 heartbeat timeouts       999999999999 - unordered
17999999999999 path MTU timeouts
18999999999999 autoclose timeouts                    SCTP user messages
19999999999999 asconf timeouts          999999999999 fragmented
20999999999999 stream reset timeouts    999999999999 reassembled
--0123456789012345678901234567890123456789012345678901234567890123456789012345
--0         1         2         3         4         5         6         7
*/

WINDOW *
opensctp(void)
{
	return (subwin(stdscr, LINES-3-1, 0, MAINWIN_ROW, 0));
}

void
closesctp(WINDOW *w)
{
	if (w != NULL) {
		wclear(w);
		wrefresh(w);
		delwin(w);
	}
}

void
labelsctp(void)
{
	wmove(wnd, 0, 0); wclrtoeol(wnd);
#define L(row, str) mvwprintw(wnd, row, 13, str)
#define R(row, str) mvwprintw(wnd, row, 51, str);
	L(0, "SCTP Associations");		R(0, "SCTP Packets");
	L(1, "associations initiated");		R(1, "packets sent");
	L(2, "associations accepted");		R(2, "packets received");
	L(3, "associations established");	R(3, "- out of the blue");
	L(4, "associations restarted");		R(4, "- bad vtag");
	L(5, "associations terminated");	R(5, "- bad crc32c");
	L(6, "associations aborted");

	L(8, "SCTP Timers");			R(8, "SCTP Chunks");
	L(9, "init timeouts");			R(9, "control chunks sent");
	L(10, "cookie timeouts");		R(10, "data chunks sent");
	L(11, "data timeouts");			R(11, "- ordered");
	L(12, "delayed sack timeouts");		R(12, "- unordered");
	L(13, "shutdown timeouts");		R(13, "control chunks received");
	L(14, "shutdown-ack timeouts");		R(14, "data chunks received");
	L(15, "shutdown guard timeouts");	R(15, "- ordered");
	L(16, "heartbeat timeouts");		R(16, "- unordered");
	L(17, "path MTU timeouts");
	L(18, "autoclose timeouts");		R(18, "SCTP User Messages");
	L(19, "asconf timeouts");		R(19, "fragmented");
	L(20, "stream reset timeouts");		R(20, "reassembled");
#undef L
#undef R
}

static void
domode(struct sctpstat *ret)
{
	const struct sctpstat *sub;
	int divisor = 1;

	switch(currentmode) {
	case display_RATE:
		sub = &oldstat;
		divisor = (delay > 1000000) ? delay / 1000000 : 1;
		break;
	case display_DELTA:
		sub = &oldstat;
		break;
	case display_SINCE:
		sub = &initstat;
		break;
	default:
		*ret = curstat;
		return;
	}
#define DO(stat) ret->stat = (curstat.stat - sub->stat) / divisor
	DO(sctps_currestab);
	DO(sctps_activeestab);
	DO(sctps_restartestab);
	DO(sctps_collisionestab);
	DO(sctps_passiveestab);
	DO(sctps_aborted);
	DO(sctps_shutdown);
	DO(sctps_outoftheblue);
	DO(sctps_checksumerrors);
	DO(sctps_outcontrolchunks);
	DO(sctps_outorderchunks);
	DO(sctps_outunorderchunks);
	DO(sctps_incontrolchunks);
	DO(sctps_inorderchunks);
	DO(sctps_inunorderchunks);
	DO(sctps_fragusrmsgs);
	DO(sctps_reasmusrmsgs);
	DO(sctps_outpackets);
	DO(sctps_inpackets);

	DO(sctps_recvpackets);
	DO(sctps_recvdatagrams);
	DO(sctps_recvpktwithdata);
	DO(sctps_recvsacks);
	DO(sctps_recvdata);
	DO(sctps_recvdupdata);
	DO(sctps_recvheartbeat);
	DO(sctps_recvheartbeatack);
	DO(sctps_recvecne);
	DO(sctps_recvauth);
	DO(sctps_recvauthmissing);
	DO(sctps_recvivalhmacid);
	DO(sctps_recvivalkeyid);
	DO(sctps_recvauthfailed);
	DO(sctps_recvexpress);
	DO(sctps_recvexpressm);
	DO(sctps_recvnocrc);
	DO(sctps_recvswcrc);
	DO(sctps_recvhwcrc);

	DO(sctps_sendpackets);
	DO(sctps_sendsacks);
	DO(sctps_senddata);
	DO(sctps_sendretransdata);
	DO(sctps_sendfastretrans);
	DO(sctps_sendmultfastretrans);
	DO(sctps_sendheartbeat);
	DO(sctps_sendecne);
	DO(sctps_sendauth);
	DO(sctps_senderrors);
	DO(sctps_sendnocrc);
	DO(sctps_sendswcrc);
	DO(sctps_sendhwcrc);

	DO(sctps_pdrpfmbox);
	DO(sctps_pdrpfehos);
	DO(sctps_pdrpmbda);
	DO(sctps_pdrpmbct);
	DO(sctps_pdrpbwrpt);
	DO(sctps_pdrpcrupt);
	DO(sctps_pdrpnedat);
	DO(sctps_pdrppdbrk);
	DO(sctps_pdrptsnnf);
	DO(sctps_pdrpdnfnd);
	DO(sctps_pdrpdiwnp);
	DO(sctps_pdrpdizrw);
	DO(sctps_pdrpbadd);
	DO(sctps_pdrpmark);

	DO(sctps_timoiterator);
	DO(sctps_timodata);
	DO(sctps_timowindowprobe);
	DO(sctps_timoinit);
	DO(sctps_timosack);
	DO(sctps_timoshutdown);
	DO(sctps_timoheartbeat);
	DO(sctps_timocookie);
	DO(sctps_timosecret);
	DO(sctps_timopathmtu);
	DO(sctps_timoshutdownack);
	DO(sctps_timoshutdownguard);
	DO(sctps_timostrmrst);
	DO(sctps_timoearlyfr);
	DO(sctps_timoasconf);
	DO(sctps_timodelprim);
	DO(sctps_timoautoclose);
	DO(sctps_timoassockill);
	DO(sctps_timoinpkill);

	DO(sctps_hdrops);
	DO(sctps_badsum);
	DO(sctps_noport);
	DO(sctps_badvtag);
	DO(sctps_badsid);
	DO(sctps_nomem);
	DO(sctps_fastretransinrtt);
	DO(sctps_markedretrans);
	DO(sctps_naglesent);
	DO(sctps_naglequeued);
	DO(sctps_maxburstqueued);
	DO(sctps_ifnomemqueued);
	DO(sctps_windowprobed);
	DO(sctps_lowlevelerr);
	DO(sctps_lowlevelerrusr);
	DO(sctps_datadropchklmt);
	DO(sctps_datadroprwnd);
	DO(sctps_ecnereducedcwnd);
	DO(sctps_vtagexpress);
	DO(sctps_vtagbogus);
	DO(sctps_primary_randry);
	DO(sctps_cmt_randry);
	DO(sctps_slowpath_sack);
	DO(sctps_wu_sacks_sent);
	DO(sctps_sends_with_flags);
	DO(sctps_sends_with_unord);
	DO(sctps_sends_with_eof);
	DO(sctps_sends_with_abort);
	DO(sctps_protocol_drain_calls);
	DO(sctps_protocol_drains_done);
	DO(sctps_read_peeks);
	DO(sctps_cached_chk);
	DO(sctps_cached_strmoq);
	DO(sctps_left_abandon);
	DO(sctps_send_burst_avoid);
	DO(sctps_send_cwnd_avoid);
	DO(sctps_fwdtsn_map_over);
	DO(sctps_queue_upd_ecne);
#undef DO
}

void
showsctp(void)
{
	struct sctpstat stats;

	memset(&stats, 0, sizeof stats);
	domode(&stats);

#define DO(stat, row, col) \
	mvwprintw(wnd, row, col, "%12lu", stats.stat)
#define	L(row, stat) DO(stat, row, 0)
#define	R(row, stat) DO(stat, row, 38)
	L(1, sctps_activeestab);	R(1, sctps_outpackets);
	L(2, sctps_passiveestab);	R(2, sctps_inpackets);
	L(3, sctps_currestab);		R(3, sctps_outoftheblue);
	L(4, sctps_restartestab);	R(4, sctps_badvtag);
	L(5, sctps_shutdown);		R(5, sctps_checksumerrors);
	L(6, sctps_aborted);


	L(9, sctps_timoinit);		R(9, sctps_outcontrolchunks);
	L(10, sctps_timocookie);	R(10, sctps_senddata);
	L(11, sctps_timodata);		R(11, sctps_outorderchunks);
	L(12, sctps_timosack);		R(12, sctps_outunorderchunks);
	L(13, sctps_timoshutdown);	R(13, sctps_incontrolchunks);
	L(14, sctps_timoshutdownack);	R(14, sctps_recvdata);
	L(15, sctps_timoshutdownguard);	R(15, sctps_inorderchunks);
	L(16, sctps_timoheartbeat);	R(16, sctps_inunorderchunks);
	L(17, sctps_timopathmtu);
	L(18, sctps_timoautoclose);
	L(19, sctps_timoasconf);	R(19, sctps_fragusrmsgs);
	L(20, sctps_timostrmrst);	R(20, sctps_reasmusrmsgs);
#undef DO
#undef L
#undef R
}

int
initsctp(void)
{
	size_t len;
	const char *name = "net.inet.sctp.stats";

	len = 0;
	if (sysctlbyname(name, NULL, &len, NULL, 0) < 0) {
		error("sysctl getting sctpstat size failed");
		return 0;
	}
	if (len > sizeof curstat) {
		error("sctpstat structure has grown--recompile systat!");
		return 0;
	}
	if (sysctlbyname(name, &initstat, &len, NULL, 0) < 0) {
		error("sysctl getting sctpstat failed");
		return 0;
	}
	oldstat = initstat;
	return 1;
}

void
resetsctp(void)
{
	size_t len;
	const char *name = "net.inet.sctp.stats";

	len = sizeof initstat;
	if (sysctlbyname(name, &initstat, &len, NULL, 0) < 0) {
		error("sysctl getting sctpstat failed");
	}
	oldstat = initstat;
}

void
fetchsctp(void)
{
	size_t len;
	const char *name = "net.inet.sctp.stats";

	oldstat = curstat;
	len = sizeof curstat;
	if (sysctlbyname(name, &curstat, &len, NULL, 0) < 0) {
		error("sysctl getting sctpstat failed");
	}
	return;
}
